### PR TITLE
[FIX] mail_message: improve perf of message_fetch_failed

### PR DESCRIPTION
**Pull request by adda-odoo**
Issue -->
`message_fetch_failed` is called when the `mail_init_messaging` controller is called every time a user navigates through Odoo

The search query made on the mail_message table to retrieve failed messages utilizes a subquery on mail_message_res_partner_needaction_rel which causes the entire query to run slow when there is a large number of rows in mail_message_res_partner_needaction_rel.

Solution -->
Replace the search call with a query that uses a join instead of a subquery and browse the returned message ids on `mail.message`

Observation -->
SQL Query returns message ids about 70% faster, hence improving perf of the method as a whole

Benchmark -->

The search call made on `mail.message` has the `has_error` field in its domain, which causes the resultant SQL query to use the `mail_message_res_partner_needaction_rel` table to filter `mail_message` via a nested query.
`SELECT * FROM mail_message WHERE id in (SELECT mail_message_id FROM mail_message_res_partner_needaction_rel WHERE ... ) ... LIMIT 100` 

For example, with a `mail_message` with  20033796 rows and  `mail_message_res_partner_needaction_rel` with 681706 rows -->

Running `explain analyze` for the author with the highest number of mail_messages on the nested query results in an execution time of 14441.994 ms.

```
 Limit  (cost=68.87..1608.20 rows=100 width=4) (actual time=39.064..14441.918 rows=37 loops=1)
   ->  Merge Semi Join  (cost=68.87..1726312.28 rows=112143 width=4) (actual time=39.062..14441.886 rows=37 loops=1)
         Merge Cond: (mail_message.id = mail_message_res_partner_needaction_rel.mail_message_id)
         ->  Index Scan Backward using mail_message_pkey on mail_message  (cost=0.44..1673763.90 rows=3336840 width=4) (actual time=0.013..13675.731 rows=3148563 loops=1)
               Filter: ((model IS NOT NULL) AND (res_id <> 0) AND (((message_type)::text <> 'user_notification'::text) OR (message_type IS NULL)) AND (author_id = 2))
               Rows Removed by Filter: 16884776
         ->  Index Scan Backward using mail_message_res_partner_needaction_rel_mail_message_id_index on mail_message_res_partner_needaction_rel  (cost=0.42..41403.42 rows=672777 width=4) (actual time=0.077..567.474 rows=672911 loops=1)
               Filter: ((notification_status)::text = ANY ('{bounce,exception}'::text[]))
               Rows Removed by Filter: 8749
 Planning Time: 0.704 ms
 Execution Time: 14441.994 ms
(11 rows)
```

Using a `JOIN` on mail_message and the needaction_rel table on `mail_message_id` with the WHERE clause filters is logically equivalent and performs faster as the query plan chooses more efficient scans of the tables.
Execution time --> 3668.712 ms

```
 Limit  (cost=0.86..1096.26 rows=100 width=4) (actual time=13.314..3668.666 rows=61 loops=1)
   ->  Nested Loop  (cost=0.86..1228410.79 rows=112143 width=4) (actual time=13.313..3668.629 rows=61 loops=1)
         ->  Index Scan Backward using mail_message_res_partner_needaction_rel_mail_message_id_index on mail_message_res_partner_needaction_rel  (cost=0.42..41403.42 rows=672777 width=4) (actual time=0.021..450.488 rows=672914 loops=1)
               Filter: ((notification_status)::text = ANY ('{bounce,exception}'::text[]))
               Rows Removed by Filter: 8749
         ->  Index Scan using mail_message_pkey on mail_message m  (cost=0.44..1.76 rows=1 width=4) (actual time=0.005..0.005 rows=0 loops=672914)
               Index Cond: (id = mail_message_res_partner_needaction_rel.mail_message_id)
               Filter: ((res_id <> 0) AND (((message_type)::text <> 'user_notification'::text) OR (message_type IS NULL)) AND (author_id = 2))
               Rows Removed by Filter: 1
 Planning Time: 0.319 ms
 Execution Time: 3668.712 ms
(11 rows)
```
This is an increase in performance by 75%.

For the author_id with the second highest count in mail_message,
`100 - (3793.788 ms/12425.190 ms)*100` = 69% increase in query performance.

opw-3649726
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Aurelienvd** [comment]
@adda-odoo Hello, I didn't review the PR yet but could you add more info to the benchmark ? Like the scanned tables' cardinality and the recordset cardinality if there is any ? Thanks !

File: addons/mail/models/mail_message.py [hunk 1]
```diff
@@ -1016,13 +1016,19 @@ def _message_format(self, fnames):
...
         errors, in the format expected by the web client."""
-        messages = self.search([
-            ('has_error', '=', True),
-            ('author_id', '=', self.env.user.partner_id.id),
-            ('res_id', '!=', 0),
-            ('model', '!=', False),
-            ('message_type', '!=', 'user_notification')
-        ], limit=100)
+        author_id = self.env.user.partner_id.id
+        query = """
```
Comment (Aurelienvd) [inline addons/mail/models/mail_message.py:1020]: You should call `self.flush` for the models/fields you are scanning in the raw SQL queries. If there are pending updates in the cache not yet flushed to the database, you could get inconsistent results.

Also, raw SQL queries bypass ir.rules so you may be missing some conditions in the WHERE clause and access restricted rows. You could use `self._search([])` to first build a Query object with the record rules applied (check odoo/odoo#143543)

File: addons/mail/models/mail_message.py [hunk 2]
```diff
@@ -1016,13 +1016,19 @@ def _message_format(self, fnames):
...
+                SELECT m.id from mail_message m
+                JOIN mail_message_res_partner_needaction_rel ON m.id = mail_message_res_partner_needaction_rel.mail_message_id
+                AND mail_message_res_partner_needaction_rel.notification_status in ('bounce','exception')
```
Comment (Aurelienvd) [inline addons/mail/models/mail_message.py:1023]: This where condition is coming from a search method on a non-stored computed field `has_error`. Turning that into a static where condition is a bit error-prone if the search method gets overriden.

**Review by Aurelienvd — COMMENTED** [review]
Left a few comments.

Looking at your query plans it seems that a compound index `mail_message(id desc, author_id)` and a partial index `mail_message_res_partner_needaction_rel(mail_message_id desc) where notification_status IN ('bounce', 'exception')` should provide an additional speedup. These are case-by-case indexes, i.e. they should be added on a single database and probably not for every database through the `init` method.

Could you also try to reduce the cardinality of both `mail_message` and `mail_message_res_partner_needaction_rel` tables and report how the query plans evolve/perform ? I know that mail_message usually has a high cardinality but I was wondering if there was any difference when there are 2M mail_message vs 20M.

**Comment by seb-odoo** [comment]
Hello, thank you for the investigation.
Note that the missing indexes have been added starting from v16 at https://github.com/odoo/odoo/pull/83451

v14 is no longer supported so this PR shouldn't be merged. Best case scenario would be to fix only in v15, but is it really worth it?

I don't think it's possible to get a satisfactory result in stable without the missing indexes? 70% reduction is not really big, it's not even an order of magnitude. 20s might be reduced to 6s but that's still really bad.

**Pull request closed without merging**